src/random.cpp

```diff
@@ -129,18 +135,41 @@ static bool GetHWRand(unsigned char* ent32) {
     return false;
 }
 
-void RandAddSeed()
+/** Use repeated SHA512 to strengthen the randomness in state. */
+static void Strengthen(CSHA512& hasher, int microseconds)
 {
-    // Seed with CPU performance counter
-    int64_t nCounter = GetPerformanceCounter();
-    RAND_add(&nCounter, sizeof(nCounter), 1.5);
-    memory_cleanse((void*)&nCounter, sizeof(nCounter));
+    static const unsigned char MARKER_OUTPUT[128] = {'O'};
+    static const unsigned char MARKER_LOCAL[128] = {'L'};
+
+    // Fork the state into a local copy, but write a distinct marker on each side
+    CSHA512 local_hasher = hasher;
+    hasher.Write(MARKER_OUTPUT, sizeof(MARKER_OUTPUT));
+    local_hasher.Write(MARKER_LOCAL, sizeof(MARKER_LOCAL));
+
+    // Hash loop
+    unsigned char buffer[64];
+    int64_t start = GetTimeMicros();
+    do {
+        for (int i = 0; i < 1000; ++i) {
+            local_hasher.Finalize(buffer);
+            local_hasher.Reset();
+            local_hasher.Write(buffer, sizeof(buffer));
+        }
+        // Write benchmark data (into output hasher)
+        int64_t perf = GetPerformanceCounter();
+        hasher.Write((const unsigned char*)&perf, sizeof(perf));
+    } while (GetTimeMicros() < start + microseconds);
```

Comment:
Maybe we should be checking for shutdown interruption in this loop?  At 10ms it isn't critical...

Reply:
Marking as resolved as the hardening code is removed from this PR.